tests/devapps/NetFxConsoleTestApp/Program.cs

```diff
@@ -309,24 +345,115 @@ 1. IWA
             Console.WriteLine("Token is {0}", authTask.Result.AccessToken);
             Console.ResetColor();
 
+            string authHeader = authTask.Result.CreateAuthorizationHeader();
 
-            await CallGraphAsync(authTask.Result.AccessToken).ConfigureAwait(false);
+            if (_usePoP)
+            {
+                await CallPoPVerificationAPIAsync(authHeader).ConfigureAwait(false);
+            }
+            else
+            {
+                await CallGraphAsync(authHeader).ConfigureAwait(false);
+            }
 
-            await CallGraphAsync(authTask.Result.AccessToken).ConfigureAwait(false);
 
             Console.BackgroundColor = ConsoleColor.DarkMagenta;
             await DisplayAccountsAsync(pca).ConfigureAwait(false);
             Console.ResetColor();
 
+        }
+
+        private static async Task CallGraphAsync(string authHeader)
+        {
+            var httpClient = new HttpClient();
+            HttpResponseMessage response;
+            var request = new System.Net.Http.HttpRequestMessage(HttpMethod.Get, GraphAPIEndpoint);
+            request.Headers.Add("Authorization", authHeader);
+            response = await httpClient.SendAsync(request).ConfigureAwait(false);
+
+            await PrintHttpResponseAsync(response).ConfigureAwait(false);
+
+        }
+
+        private static async Task PrintHttpResponseAsync(HttpResponseMessage response)
+        {
+            var content = await response.Content.ReadAsStringAsync().ConfigureAwait(false);
+
+            if (response.IsSuccessStatusCode)
+            {
+                Console.BackgroundColor = ConsoleColor.DarkGreen;
+                Console.WriteLine(response.StatusCode);
+                Console.WriteLine(content);
+
+                Console.ResetColor();
+            }
+            else
+            {
+                Console.BackgroundColor = ConsoleColor.DarkRed;
+                Console.ForegroundColor = ConsoleColor.White;
+                Console.WriteLine(response.StatusCode);
+                Console.WriteLine(content);
+
+                Console.ResetColor();
+            }
+
+        }
+
+        /// <summary>
+        /// This calls a special endpoint that validates any POP token against a configurable http request.
+        /// The http request is configured through headers.
+        /// </summary>
+        private static async Task CallPoPVerificationAPIAsync(string authHeader)
+        {
+            var httpClient = new HttpClient();
+            HttpResponseMessage response;
+            var request = new HttpRequestMessage(HttpMethod.Post, PoPValidatorEndpoint);
+
+            request.Headers.Add("Authorization", authHeader);
+            request.Headers.Add("Secret", s_secretForPoPValidationRequest);
+            request.Headers.Add("Authority", "https://sts.windows.net/72f988bf-86f1-41af-91ab-2d7cd011db47/");
+            request.Headers.Add("ClientId", s_clientIdForPublicApp);
+
```

Comment:
why is all this needed? is this something developers would need to write as well in order to use PoP?

Reply:
No, this is because the SHR (George's test validator) is flexible and allows us to test any kind of PoP token. 

Users would not need to set any of this up./

Reply:
i see... very cool

---
In reply to: [352758457](https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/pull/1525#discussion_r352758457) [](ancestors = 352758457)